pkg/bootstrap/platform/gcp.go

```diff
@@ -211,3 +228,75 @@ func (e *gcpEnv) Locality() *core.Locality {
 
 	return &l
 }
+
+var (
+	mdOnce   sync.Once
+	project  string
+	zone     string
+	instance string
+	cluster  string
+)
+
+// Retrieves needed metadata strictly from the Google Compute API
+func getMetadataAttributes(e *gcpEnv) {
+	mdOnce.Do(func() {
+		project, _ = e.projectIDFn()
+		zone, _ = e.locationFn()
+		instance, _ = e.instanceNameFn()
+		cluster, _ = e.clusterNameFn()
+	})
+}
+
+// Labels attempts to retrieve the GCE instance labels within the timeout
+// Requires read access to the Compute API (compute.instances.get)
+func (e *gcpEnv) Labels() map[string]string {
+	getMetadataAttributes(e)
+	ctx, cancel := context.WithTimeout(context.Background(), defaultTimeout)
+	defer cancel()
+
+	success := make(chan bool)
+	labels := map[string]string{}
+	var instanceLabels map[string]string
+	go func() {
+		// use explicit credentials with compute.instances.get IAM permissions
+		creds, err := google.FindDefaultCredentials(ctx, compute.CloudPlatformScope)
+		if err != nil {
+			log.Warnf("failed to find default credentials: %v", err)
+			success <- false
+			return
+		}
+		computeService, err := compute.NewService(ctx, option.WithCredentials(creds))
+		if err != nil {
+			log.Warnf("failed to create new service: %v", err)
+			success <- false
+			return
+		}
+		// instance.Labels is nil if no labels are present
+		instanceObj, err := computeService.Instances.Get(project, zone, instance).Do()
+		if err != nil {
+			log.Warnf("unable to retrieve instance: %v", err)
+			success <- false
+			return
+		}
+		instanceLabels = instanceObj.Labels
+		success <- true
+	}()
+	select {
+	case <-ctx.Done():
+		log.Warnf("context deadline exceeded for instance get request: %v", ctx.Err())
+	case ok := <-success:
+		if ok && instanceLabels != nil {
+			labels = instanceLabels
+		}
+	}
+	return labels
+}
+
+const KubernetesServiceHost = "KUBERNETES_SERVICE_HOST"
```

Comment:
nit: it is fine to have this here, but you might consider moving it to the `const` block at the top of this file.